README.md

```diff
@@ -1,138 +1,615 @@
+  - Note, there is no third argument; the property descriptor is not passed into the callback, but the callback could look it up itself.
+- The return value of the callback must be undefined.
+- All `@wrap` callbacks run before all `@register` callbacks. This is because `@wrap` is used to set up the class, and `@register` runs on the class after it's created.
 
-## Private methods and fields
+### `@initialize`
 
-The above example has some implementation details exposed to the world that might be better kept internal. Using ESnext [private fields and methods](https://github.com/tc39/proposal-private-methods), the definition can be refined to:
+The `@initialize` decorator intercepts the initialization of a public class field and runs a callback supplied to the decorator in place of `Object.defineProperty`. For example:
 
 ```js
-class Counter extends HTMLElement {
-  #x = 0;
+class C {
+  @initialize(f) a = b;
+}
+```
 
-  #clicked() {
-    this.#x++;
-    window.requestAnimationFrame(this.render.bind(this));
-  }
+is roughly equivalent to the following:
 
+```js
+class C {
   constructor() {
-    super();
-    this.onclick = this.#clicked.bind(this);
+    f.call(this, f, b, "a");
   }
+}
+```
 
-  connectedCallback() { this.render(); }
+Details:
+- `@initialize` can only be used on public field declarations (including symbols/computed property names), not private.
+- `@initialize` can be used with both static and instance fields.
+- The return value of the callback must be undefined.
 
-  render() {
-    this.textContent = this.#x.toString();
-  }
+## User-defined Decorators
+
+JavaScript programmers can make thier own decorators by composing built-in decorators.
+
+### `decorator @xyz` declarations
+
+Decorators may be defined as a simple composition of other decorators. You can use all the fancy JavaScript features you want inside the arguments, but at the top level, this is just a string of decorators and arguments for these decorators. There's no way to conditionally use one decorator in one situation and another in another situation, for example.
+
+Example:
+
+```js
+decorator @xyz {
+  @foo @bar(arg) @baz(arg2)
 }
-window.customElements.define('num-counter', Counter);
+@xyz class C { }
 ```
 
-To make methods and fields private, just give them a name starting with `#`. A shorthand for `this.#x` is simply `#x`.
+This is basically equivalent to listing those decorators explicitly:
 
-By defining things which are not visible outside of the class, ESnext provides stronger encapsulation, ensuring that your classes' users don't accidentally trip themselves up by depending on internals, which may change version to version.
+```js
+@foo @bar(arg) @baz(arg2)
+class C { }
+```
 
-Note that ESnext provides private fields only as declared up-front in a field declaration; private fields cannot be created as expandos.
+Decorators may also take arguments:
 
-## Decorators
+```js
+decorator @xyz(arg, arg2) {
+  @foo @bar(arg) @baz(arg2)
```

Comment:
what if, in the absence of do expressions, i want to use *statements* to create new arguments, based on the arguments to `xyz`? How would I achieve this?

Reply:
You could factor out those statements into a function, and call the function.

You criticized a separate idea from @nicolo-ribaudo to allow statements in another thread. Do you have any ideas here?

Reply:
Maybe instead of a decorator block consisting of nothing but decorator usages, something like this?:
```js
decorator @xyz(a, b) {
  statement;
  list;
  with;
  normal;
  block;
  semantics;
  @one @plus @decorators @must @be @the @last @items @in @the @block?
}
```

Reply:
OK, let's follow up in #252 .